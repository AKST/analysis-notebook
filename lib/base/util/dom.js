/**
 * @param {Node} parent
 * @param {Node} child
 */
export function insertAtStart(parent, child) {
  if (parent.firstChild) {
    parent.insertBefore(child, parent.firstChild);
  } else {
    parent.appendChild(child);
  }
}

/**
 * @param {HTMLElement | undefined} element
 * @param {number | boolean | string | null | undefined} value
 */
export function setDisable(element, value) {
  if (element == null) return;
  switch (value) {
    case undefined:
    case null:
    case '':
    case false:
    case 0:
      element.removeAttribute('disabled');
      break;
    default:
      element.setAttribute('disabled', 'disabled');
  }
}
