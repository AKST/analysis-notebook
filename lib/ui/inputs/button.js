/**
 * @import { El } from '../../base/dom_ui/index.js';
 */
import { generateCSS } from '../../base/dom_ui/css.js';
import { render, updateOn } from '../../base/dom_ui/index.js';

/**
 * @typedef {Object} ButtonAppearance
 * @property {'destructive'} [action]
 * @property {'thin'} [padding]
 */

export class BasicButton extends HTMLElement {
  /** @type {ButtonAppearance} */
  appearance;

  /** @type {HTMLButtonElement | null} */
  button = null;

  /**
   * @param {El} content
   * @param {ButtonAppearance} [appearance={}]
   */
  constructor(content, appearance = {}) {
    super();
    this.attachShadow({ mode: 'open' });
    this.content = content;
    this.appearance = appearance
  }

  connectedCallback() {
    const button = this.button = document.createElement('button');
    this.button.className = 'button';
    this.render();
    // this.addEventListener('keydown', e => {
    //   if (e.key !== 'Enter') return;
    //   this.button.dispatchEvent(new CustomEvent('click'));
    // });

    const eventsToForward = [
      'click', 'mousedown', 'mouseup', 'mouseover', 'mouseout',
      'mouseenter', 'mouseleave', 'focus', 'blur', 'keydown', 'keyup'
    ];

    eventsToForward.forEach(eventType => {
      button.addEventListener(eventType, (/** @type {any} */ e) => {
        const newEvent = new e.constructor(e.type, {
          bubbles: e.bubbles,
          cancelable: e.cancelable,
          detail: e.detail,
          ...e
        });
        e.stopPropagation();
        this.dispatchEvent(newEvent);
      });
    });

  }

  focus() {
    this.button?.focus()
  }

  render() {
    this.button = /** @type {HTMLButtonElement} */
      (render(['button', { class: 'button', tabIndex: '0' }, [
        this.content ?? ['slot', null, []]
      ]]));

    updateOn(this.shadowRoot, [
      ['style', null, [getStyle(this.appearance)]],
      this.button,
    ])
  }
}

customElements.define('basic-button', BasicButton);

/**
 * @param {ButtonAppearance} param0
 */
function getStyle({ action, padding }) {
  const bg = { destructive: 'red' };
  const fg = { destructive: 'white' };
  return generateCSS({
    '.button': {
      display: 'block',
      width: '100%',
      color: (action && fg[action]) ?? 'black',
      background: (action && bg[action]) ?? 'buttonface',
      border: 'var(--border-outset-grey)',
      borderWidth: 2,
      padding: padding === 'thin' ? [2, 4, 2] : [2, 8, 2],
      '&:active': {
        borderStyle: 'inset',
      },
    },
  });
}
