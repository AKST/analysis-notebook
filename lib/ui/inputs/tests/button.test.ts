import { expect, describe, it, beforeEach, vi } from 'vitest';
import userEvent from '@testing-library/user-event';
import { getByRole, getByText } from '@testing-library/dom';
import { BasicButton } from '../button.js';

describe('BasicButton', () => {
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    document.body.innerHTML = '';
    user = userEvent.setup();
  });

  it('renders default button correctly', () => {
    const button = new BasicButton('Click me');
    document.body.appendChild(button);
    expect(button.shadowRoot?.innerHTML).toMatchSnapshot();
  });

  it('renders destructive button correctly', () => {
    const button = new BasicButton('Delete', { action: 'destructive' });
    document.body.appendChild(button);
    expect(button.shadowRoot?.innerHTML).toMatchSnapshot();
  });

  it('responds to user clicks', async () => {
    const clickHandler = vi.fn();
    const button = new BasicButton('Click me');
    button.addEventListener('click', clickHandler);
    document.body.appendChild(button);
    await user.click(button);
    expect(clickHandler).toHaveBeenCalledOnce();
  });
});
