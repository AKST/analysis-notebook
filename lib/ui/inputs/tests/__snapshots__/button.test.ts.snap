// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BasicButton > renders default button correctly 1`] = `
"<style>.button {
  display: block;
  width: 100%;
  color: black;
  background: buttonface;
  border: var(--border-outset-grey);
  border-width: 2px;
  padding: 2px 8px 2px;
  &:active {
    border-style: inset;
  }

}

</style><button class="button" tabindex="0">Click me</button>"
`;

exports[`BasicButton > renders destructive button correctly 1`] = `
"<style>.button {
  display: block;
  width: 100%;
  color: white;
  background: red;
  border: var(--border-outset-grey);
  border-width: 2px;
  padding: 2px 8px 2px;
  &:active {
    border-style: inset;
  }

}

</style><button class="button" tabindex="0">Delete</button>"
`;
